applications/nrf_desktop/src/modules/ble_qos.c

```diff
@@ -0,0 +1,689 @@
+		static u32_t m_cdc_notify_count;
+
+		/* Count number of HID packets received via BLE */
+		/* and number of HID packets sent via USB. */
+		/* Send stats printout via CDC every 100 packets */
+
+		if (is_hid_mouse_event(eh)) {
+			++m_hid_pkt_recv_count;
+			++m_cdc_notify_count;
+
+			if (m_cdc_notify_count == 100) {
+				hid_pkt_stats_print(
+					m_hid_pkt_recv_count,
+					m_usb_pkt_sent_count);
+				m_cdc_notify_count = 0;
+			}
+
+			return false;
+		}
+
+		if (is_hid_report_sent_event(eh)) {
+			++m_usb_pkt_sent_count;
+
+			return false;
+		}
+	}
+#endif
+
+	/* If event is unhandled, unsubscribe. */
+	__ASSERT_NO_MSG(false);
+
+	return false;
+}
+
+static void apply_new_params(void)
+{
+	int err;
+
+	k_mutex_lock(&m_data_access_mutex, K_FOREVER);
+	/* chmap_filter_params_set Returns immediately */
+	err = chmap_filter_params_set(mp_chmap_inst, m_new_params);
+	k_free(m_new_params);
+	m_new_params = NULL;
+	k_mutex_unlock(&m_data_access_mutex);
+
+	if (err) {
+		LOG_WRN("Param update failed");
+	}
+}
+
+void ble_qos_thread_fn(void)
+{
+	int err;
+
+	err = k_sem_take(&sem_initialized, K_FOREVER);
+	if (err) {
+		LOG_ERR("Init failed: %d", err);
+		__ASSERT_NO_MSG(false);
+	}
+
+	for (;;) {
+		bool update_channel_map;
+
+		k_sleep(1000); /* Set time unit for chmap filter to ~1 second */
+
+		/* Check and apply new parameters received via config channel */
+		if (m_new_params != NULL) {
+			apply_new_params();
+		}
+
+		/* Check and apply new blacklist received via config channel */
+		atomic_t blacklist_update = atomic_set(&m_new_blacklist, -1);
+
+		if (blacklist_update != -1) {
+			err = chmap_filter_blacklist_set(
+						mp_chmap_inst,
+						(u16_t) blacklist_update);
+			if (err) {
+				LOG_WRN("Blacklist update failed");
+			}
+		}
+
+		/* Run processing function. */
+		/* Atomic variable is used as data busy flag */
+		/* (this thread runs at the lowest priority) */
+		atomic_set(&m_processing, 1);
+		update_channel_map = chmap_filter_process(mp_chmap_inst);
+		atomic_set(&m_processing, 0);
+
+#if CONFIG_DESKTOP_BLE_QOS_STATS_PRINTOUT_ENABLE
+		ble_chn_stats_print(update_channel_map);
+#endif
+
+		if (!update_channel_map) {
+			continue;
+		}
+
+		u8_t const *chmap;
+
+		chmap = chmap_filter_suggested_map_get(mp_chmap_inst);
+		err = hci_chmap_cmd_send(chmap);
+		if (err) {
+			LOG_WRN("hci_chmap_cmd_send: %d", err);
+		} else {
+			LOG_DBG("Channel map update");
+			chmap_filter_suggested_map_confirm(mp_chmap_inst);
+
+			k_mutex_lock(&m_data_access_mutex, K_FOREVER);
+			memcpy(m_current_chmap, chmap, sizeof(m_current_chmap));
+			k_mutex_unlock(&m_data_access_mutex);
+		}
+	}
+}
+
+#if CONFIG_DESKTOP_BLE_QOS_STATS_PRINTOUT_ENABLE
+#define STACKSIZE 1024
+#else
+#define STACKSIZE 240
+#endif
+/* Lowest thread priority */
+#define PRIORITY K_PRIO_PREEMPT((CONFIG_NUM_PREEMPT_PRIORITIES - 1))
+
+K_THREAD_DEFINE(ble_qos_thread_id, STACKSIZE, ble_qos_thread_fn, NULL, NULL,
+		NULL, PRIORITY, 0, K_NO_WAIT);
```

Comment:
you can try putting K_WAITFOREVER and k_thread_wakeup after initialization is done. Or simply create there there